File: cs/1_5_prototype.cs
using System;

namespace Creational
{
    namespace Prototype
    {
        interface IPrototype {
            IPrototype Clone();
        }
        class SomeType: IPrototype
        {
            public string Name = "No name";
            public int Count = 0;

            public IPrototype Clone()
            {
                return this.MemberwiseClone() as SomeType;
            }
        }
        class ProductPrototype: IPrototype
        {
            protected DateTime CreatedAt;
            protected int Id;

            public ProductPrototype() 
            {
                this.CreatedAt = DateTime.Now;
                Random rnd = new Random();
                this.Id = rnd.Next();
            }

            public virtual IPrototype Clone()
            {
                return this.MemberwiseClone() as ProductPrototype;
            }
        }

        class CustomProduct : ProductPrototype
        {
            public SomeType obj;
            public CustomProduct(SomeType obj = null) : base()
            {
                this.obj = obj;
            }
            public override CustomProduct Clone()
            {
                CustomProduct clone = this.MemberwiseClone() as CustomProduct;
                clone.obj = this.obj.Clone() as SomeType;
                clone.CreatedAt = DateTime.Now;
                return clone;
                // return this.MemberwiseClone() as CustomProduct; // Deep copy problem
            }

            public override string ToString()
            {
                return $"{{\n\tId:{this.Id},\n\tCreatedAt:{this.CreatedAt},\n\tobj:{{ {(this.obj as SomeType).Name} }}\n}}";
            }
        }
    }
}

File: js/1_5_prototype.js
class Prototype{
    clone(){
        return Object.assign({}, this);
    }
}

class SomeType {
    constructor() {
        this.name = "Noname";
        this.count = 0;
    }
}

class ProductPrototype extends Prototype {
    constructor() {
        this.createdAt = new Date(Date.now());
        this.id = Math.random();
    }
}

class CustomProduct extends ProductPrototype {
    constructor(obj) {
        super();
        this.obj = obj;
    }
    clone() {
        let clone = super.clone();
        clone.obj = Object.assign({}, this.obj);
        clone.createdAt = new Date(Date.now());
        return clone;
    }
}

export { SomeType, ProductPrototype, CustomProduct };